src/librustc_lint/builtin.rs

```diff
@@ -1927,8 +1932,23 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for InvalidValue {
                             return Some(InitKind::Zeroed);
                         }
                     }
-                    // FIXME: Also detect `MaybeUninit::zeroed().assume_init()` and
-                    // `MaybeUninit::uninit().assume_init()`.
+                }
+            } else if let hir::ExprKind::MethodCall(_, _, ref args) = expr.kind {
+                // Find problematic calls to `MaybeUninit::assume_init`.
+                let def_id = cx.tables.type_dependent_def_id(expr.hir_id)?;
+                if cx.tcx.is_diagnostic_item(sym::assume_init, def_id) {
+                    // This is a call to *some* method named `assume_init`.
+                    // See if the `self` parameter is one of the dangerous constructors.
+                    if let hir::ExprKind::Call(ref path_expr, _) = args[0].kind {
+                        if let hir::ExprKind::Path(ref qpath) = path_expr.kind {
+                            let def_id = cx.tables.qpath_res(qpath, path_expr.hir_id).opt_def_id()?;
+                            if cx.match_def_path(def_id, MU_ZEROED_PATH) {
```

Comment:
Shouldn't this (and the other one below) also use diagnostics items?

Reply:
Yes, see https://github.com/rust-lang/rust/issues/66075. This is partially pre-existing. I didn't feel like rewriting the entire lint at once.^^

Reply:
yea we should be moving away from `match_def_path` where we can. That's the reason diagnostic items were introduced

Reply:
@RalfJung Sure that's fair about the old code... but these are additions?

Reply:
But I guess this is as good as any a time to resolve the first bullet point from that issue. Not sure when I'll get to it, though.